### ISSUE TITLE
[JOSS review] Article comments

### ISSUE THREAD

**Issue by nbeliy** [openneuropet/PET2BIDS#236]
Dear pet2bids authors, this is in reference to the review of the paper that I'm currently conducting for JOSS [here](https://github.com/openjournals/joss-reviews/issues/6067).

First thanks to put an effort for developing the bidsifier for PET data -- the BIDS specification for PET is quite complex, and your tool is for sure is needed!

I will put my comments in the checklist below:

### General comments

- [ ] `Statement of needs` do not state the need of software, you may add a small trivial phrase why pet2bids exists (not what it does).
- [ ] JOSS asks to compare the software with similar tools
- [ ] `File conversation`/`PET Metadata`/`Spreadsheet conversation` may be worth to be put in the dedicated (sub)sections
- [ ] Are conversation, metadata and spreadsheets export are extensive list of pet2bids tools? 
- [ ] In Metadata/Spreadsheet section you need to change `JSON file` to `sidecar JSON file`, first one is an any generic JSON file, second is BIDS specific one that supplements image data with metadata.
- [ ] In spreadsheets section, I don't really like the usage of points (`(a)` ,  `(i)`, etc.) -- you never reference them. Maybe replace the first `and` at line 54 by `as well as`, and make the iteration in lines 55-58 to a list. I think it would be easier to read.

### Inline comments

- [ ] On line 27, missing references to ECAT and DICOM formats
- [ ] On line 27, by time measurements do you mean the timing of each PET frame? Just to be sure.
- [ ] Line 30 is a little bit confusing, do you mean by `library code, allowing conversion ... using the command line`, that you provide a library **and** a (command line) script?
- [ ] Line 30, I may be wrong here, but `library code` reads as a code that do something with library. Library of tools/functions sounds better for me.
- [ ]  Line 31, `it can be integrated into software` because it's library, not because it modular. Need to be rephrased.
- [ ]  Line 39, needed references for BT Christian and RF Muzic, are those people or tools or functions?
- [ ]  Line 43, What do you mean by `the further testing of data reading`?
- [ ]  Line 46, It's not immediately clear that `JSON files created from reading PET` refers to JSON sidecar files from conversion using pet2bids.
- [ ]  Line 46, `data are always missing...`, `always` seems too strong, maybe better to replace by `often` or `usually`.
- [ ] Line 48, as original PET data (DICOM/ECAT) don't contain JSON, there no `original JSON file`, I guess you meant sidecar JSON file?
- [ ] Line 48, before you provided the names of function (for ex. ecat2nii for conversion), but here you use `PET JSON updater`, can you change it into function name(s)?
- [ ] Line 52, is `xls, xlsx, csv, tsv, bld` it extensive list of tabular data formats in existence, or list of data  that bids2pet supports? If it's first, than an `for example` may be needed.
- [ ] Line 57, what is `blood.tsv` file, you never mentioned it before. Maybe rephrase it to `exports blood data from PMOD tabular file`?

**Comment by CPernet** [comment]
ready to go once #255 is fixed

**Comment by CPernet** [comment]
- [ ] `Statement of needs` do not state the need of software, you may add a small trivial phrase why pet2bids exists (not what it does).

The guidelines https://joss.readthedocs.io/en/latest/submitting.html say 'A Statement of need section that clearly illustrates the research purpose of the software and places it in the context of related work'. Therefore, what it does is required. We have added however the 'why' we created it as introductory sentence.

- [ ] JOSS asks to compare the software with similar tools

There are no other tools doing that type of conversion for PET data, and mention that in the 'why' introductory sentence.

- [ ] `File conversation`/`PET Metadata`/`Spreadsheet conversation` may be worth to be put in the dedicated (sub)sections

We could but rather stick to the minimalistic recommended JOSS guideline section format.

- [ ] Are conversation, metadata and spreadsheets export are extensive list of pet2bids tools? 

The section on conversion was rewritten to make it clear is it done using a single wrapper function for DICOM files or a single wrapper function for Ecat files. We also clarified that there are 3 spreadsheets export functions.

- [ ] In Metadata/Spreadsheet section you need to change `JSON file` to `sidecar JSON file`, first one is an any generic JSON file, second is BIDS specific one that supplements image data with metadata.

JSON are always sidecar files, we now distinguish the image from tsv sidecar JSON.

- [ ] In spreadsheets section, I don't really like the usage of points (`(a)` ,  `(i)`, etc.) -- you never reference them. Maybe replace the first `and` at line 54 by `as well as`, and make the iteration in lines 55-58 to a list. I think it would be easier to read.

Given that we now explain that there are 3 functions, those points are logical - but removed a and b.

### Inline comments

- [ ] On line 27, missing references to ECAT and DICOM formats

added ref to DICOM and added working for ECAT is proprietary

- [ ] On line 27, by time measurements do you mean the timing of each PET frame? Just to be sure.

not necessarily, frame information is ofen present, so we added 'e.g. time zero' as a different example. 

- [ ] Line 30 is a little bit confusing, do you mean by `library code, allowing conversion ... using the command line`, that you provide a library **and** a (command line) script?

This sentence has been removed.

- [ ] Line 30, I may be wrong here, but `library code` reads as a code that do something with library. Library of tools/functions sounds better for me.

We now use code library throughout.

- [ ]  Line 31, `it can be integrated into software` because it's library, not because it modular. Need to be rephrased.

Sure - rewritten. 

- [ ]  Line 39, needed references for BT Christian and RF Muzic, are those people or tools or functions?

Those are the people who wrote the ecat reading function.

- [ ]  Line 43, What do you mean by `the further testing of data reading`?

We meant that we tested the reading/writting of the data, i.e. which bits are read according to the PET data frames. Referering now explicitly to [ecat validation](https://github.com/openneuropet/PET2BIDS/tree/main/ecat_validation).

- [ ]  Line 46, It's not immediately clear that `JSON files created from reading PET` refers to JSON sidecar files from conversion using pet2bids.

fixed

- [ ]  Line 46, `data are always missing...`, `always` seems too strong, maybe better to replace by `often` or `usually`.

no 'always' is correct - we have never seen a JSON file with all the metadata simply because information from the scanner do not encode eg pharmaceutical information.

- [ ] Line 48, as original PET data (DICOM/ECAT) don't contain JSON, there no `original JSON file`, I guess you meant sidecar JSON file?

fixed

- [ ] Line 48, before you provided the names of function (for ex. ecat2nii for conversion), but here you use `PET JSON updater`, can you change it into function name(s)?

fixed

- [ ] Line 52, is `xls, xlsx, csv, tsv, bld` it extensive list of tabular data formats in existence, or list of data  that bids2pet supports? If it's first, than an `for example` may be needed.

changes to Supported formats are .xls, .xlsx, .csv, .tsv and .bld.

- [ ] Line 57, what is `blood.tsv` file, you never mentioned it before. Maybe rephrase it to `exports blood data from PMOD tabular file`?

The conversion goes the other way around. We have now changed this to BIDS blood.tsv to make it clear.

**Comment by CPernet** [comment]
updated are per other reviewer request as well

**Comment by nbeliy** [comment]
I feel that the structure of the article can be better.

I would suggest to add a section `Description of functionality`, which you can start with saying what exactly pet2bids do (conversion, organization, metadata extension and blood export). Then the paragraphs with technical details will be placed in it's context.

**Comment by CPernet** [comment]
that is not the JOSS format - I do not want to change their structure

**Comment by nbeliy** [comment]
Paragraph on spreadsheet conversion is still quite heavy, you can rewrite it in much more concise way, espetially that there very little difference between (i) and (ii). One just create JSON, other updates JSON. You can say that pet2bids allow to extract blood data from pre-formatted tabular data (including PMOD one) and supplement the sidecar JSON together with blood.tsv.

**Comment by nbeliy** [comment]
Are all technical details in file conversion really needed? As reader, I don't really care that you use dcm2niix or Nibabel. I feel that it can be replaced by how and what kind of metadata user can and must supply.

**Comment by CPernet** [comment]
'Are all technical details in file conversion really needed? As reader, I don't really care that you use dcm2niix or Nibabel. I feel that it can be replaced by how and what kind of metadata user can and must supply.'

it does, because we are talking quantitative data - not relative like MRI - that's why we have test counting the rounding errors and stuff like that

**Comment by CPernet** [comment]
'Paragraph on spreadsheet conversion is still quite heavy, you can rewrite it in much more concise way, espetially that there very little difference between (i) and (ii). One just create JSON, other updates JSON. You can say that pet2bids allow to extract blood data from pre-formatted tabular data (including PMOD one) and supplement the sidecar JSON together with blood.tsv.'

sure - done https://github.com/openneuropet/PET2BIDS/commit/3e28b32178575ae9e9b218e86991744fcd26090f#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6

**Comment by CPernet** [comment]
but you convert speadsheet to json, DICOM and ecst to Nifti

**Comment by nbeliy** [comment]
The starting phrase of PET metadata is still awkward. `Some ... information are always missing` sounds like a truism. You can rephrase it into something like: Missing medata from conversions (e.g. radiotracer and pharmaceutical information) can be completed by PET JSON updater.

The important point is that JSON updator updates json and check for consistency and missing information, everything else is not important.

**Comment by CPernet** [comment]
Some ... information are always missing sounds like a truis ?? ok you last comment was tat surely is not always 
I do not want to change that -- it has to be stated

**Comment by CPernet** [comment]
I would mention the CLI only at the end paragraph of L30.

I could --> 'Note that the python tools are command line tools, i.e. they can be called directly from a terminal. ' ?

**Comment by nbeliy** [comment]
> that is not the JOSS format - I do not want to change their structure

If you want, but you still need a paragraph providing an overview of what exactly pet2bids do. It may be short, but without it, `File conversion`, `Metadata` and `spreadsheet` looks out of place and technical. 

It will also resolve the issue that you never mention that pet2bids sort dataset into BIDS folder structure.

P.S. There was a way to start a tread to keep stuff organized. Can't find it( If you know how to do it, please help

**Comment by CPernet** [comment]
overview of what exactly pet2bids do.  -- that what the statement of need does as per JOSSS structure -- 'Here we present a new code library, developed in both Matlab and Python, allowing the conversion of DICOM [@dicm_ref_2020] and ECAT (CTI/Siemens proprietary data format) PET imaging data and metadata (e.g., timing information such as e.g. 'time zero' or blood measurements) into the BIDS specification.'

**Comment by CPernet** [comment]
pet2bids sort dataset into BIDS folder structure. ?? but it does not

**Comment by nbeliy** [comment]
> it does, because we are talking quantitative data - not relative like MRI - that's why we have test counting the rounding errors and stuff like that - so for PET people it is essential IMO

I understand the importance of testing. But, why do you not test conversion from DICOM, or conversion from ECAT in matlab?

It's not the problem that you mention the tests, but how you do it.

**Comment by CPernet** [comment]
I understand the importance of testing. But, why do you not test conversion from DICOM, or conversion from ECAT in matlab? 

what ?? conversion form DICOMN is tested by dcm2niix - conversion from ecat is tested both matlab and python - I do not get your point

**Comment by CPernet** [comment]
- [x]  Line 31: is the first tool to convert --> is the first tool, in our knowledge, to organize. I know at least few tools that can organize PET dataset at least to some extent. Either go with First PET dedicated, or stress that your tool do organization in the most complete way.

That's right, some tools can organize - none can convert

- [x]  Line 37, 41: If both in python and matlab, functions shares the same name, you don't need to precise extensions. Especially if in python, .py is a script, not a function.

fine

- [x]  Line 36-39: Phrase too long, difficult to read. On top of it, functions which are wrapper functions is ugly.

slightly updated, left wrapper - your opinion the name is 'ugly' 

- [x]  Line 38: That as it is placed here, refers to dcm2niix, not to dcm2niix4pet. Also it is not clear how the details not from source images are retrieved, are they supplied by user?

added

- [x] Line 40: the newly created contradicts their reliance on code of Christian and Muzic

I don't think so, the new converter relies on a function that reads

- [x] Line 42: If you can't provide reference to library from which readECAT7 was taken, move Christian and Muzic to acknowledgments. And make their names less DJ, for ex. B. T. Christian.

why? there is no convention for that in the JOSS ref system

- [x] Line 44: I would omit testing part. First it implies that matlab functions are untested, second, testing is part of unit tests, not of description of a specific function.

requested by over reviewer

- [x] Line 47: change and  must provide to with user provided

?? doesn't work as a sentence 

- [x] Line 50: Why the update_json_pet_file name is different in matlab and in python?

idiosyncrasies of the code

- [x] Line 54: The mention that dcm2niix4pet make a call to update_json_pet_file is pointless. You already stated that user can provide additional metadata to conversion, and the fact that you describe update_json_pet_file means that user can call it.

fixed

- [x] Line 62: The point (iii) implies that spreadsheet conversion extracts data for blood.tsv onlyfrom PMOD file? Why it can't extract blood information from tabular data?

PMOD is a very specific tabular type - what can it not do from others? because there is no standard for that

- [x] Line 62: Missing PMOD reference.
 
no it's there  - @Burger1997

------------
https://github.com/openneuropet/PET2BIDS/commit/ac4a7354f5756e2157f45ec1e80b4ded6270640a#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6
-------------

**Comment by nbeliy** [comment]
> pet2bids sort dataset into BIDS folder structure. ?? but it does not (hence not 'mentioned')

Exactly the point. BIDS imposes folder structure, specific naming, specific metadata and specific format. When you tell that you convert to BIDS, it is understood that you do everything. If you do only format and metadata, it must be told explicitly. And probably motivate why you don't do folder structure (which is the easiest part of bidsification).

For example you can say that the goal of your library is to manage the complex BIDS-PET metadata, and it is supposed to be used from inside of other bidsifiers like ezBIDS and BIDSCOIN.

**Comment by CPernet** [comment]
got you now, makes sense --> https://github.com/openneuropet/PET2BIDS/commit/c026ad49ebf187211a2ddeb2ec65053901ba8c96#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6

**Comment by nbeliy** [comment]
> what ?? conversion form DICOM is tested by dcm2niix - conversion from ecat is tested both matlab and python - I do not get your point

It's not what you write, it's how you write. 

`The Python code was developed to mirror the Matlab code, with further testing of data reading`. Read literally, it tells 2 things: python code (ecat2nii) contains the test of data reading, and Matlab do not contain this test. It also imply that dcm2niix4pet do not contain such tests.

I'm still not sure what exactly you wanted to say, but  your point is that because ecat2nii is home made function, we did additional tests to ensure the quality of conversion. It could look something like:

In order to insure the quality of conversion (especially important in quantitative imagery, like PET), we developed a python tool <name> that compares the original ECAT image with converted one.

**Comment by CPernet** [comment]
again, got you know -- again makes sense! 
https://github.com/openneuropet/PET2BIDS/commit/6a9c530427358543ff5060c2f1f5a8724f5f09a1#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6

**Comment by CPernet** [comment]
latest requests https://github.com/openneuropet/PET2BIDS/commit/8e3953020fb0cf1c046cf32bee0e79f0fde43fcb#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6

**Comment by CPernet** [comment]
1 - blood data are unusual and do not come with all imaging 
2 - the image json does not contain blood information -- this is optional additional metadata
3 - you were right (reworded) we propose a preformated table, or PMOD (https://github.com/openneuropet/PET2BIDS/tree/main/spreadsheet_conversion/blood)
3 - PMOD is not one file but several files to be combined

**Comment by nbeliy** [comment]
Based on draft of [c8eaf08](https://github.com/openneuropet/PET2BIDS/commit/c8eaf08b55edb8b6522732a1b5b90d6039ed8888).

I collected here the points that paper is either misleading or straight incorrect.

Line 31: 
> PET2BIDS is the first tool to convert ... to BIDS

Misleading, implying that you convert  and structure the data following BIDS. I asked several persons and they all reads it this way. To rephrase.

Line 32: 
> conversion done using command line **only**

Factually incorrect. You provide CLI only for Python, for matlab you provide functions.

Line 41:
> with details that user **must** provide **because ** ...

Factually incorrect. The functions `dcm2niix4pet` allow user to not provide all required details. To rephrase.

Line 46:
> further testing data reading...

Incomprehensible. To rewrite.

Line 55:
> The PET JSON updater functions (... update_json_pet.py)

Factually incorrect. `.py` are not functions, but files. It can contain (and actually it does) several functions. To correct.

**Comment by CPernet** [comment]
[x] 'PET2BIDS is the first tool to convert ... to BIDS'  --> made the statement valid  --> .. convert to BIDS files.  (note we already state above that it's only files and not structure)

[x] command line only -- removed only
 
[x] because -- made it correct --> 'to be BIDS valid'

[x] functions - yeah, after so many changes requested, the whole paragraph lost its meaning -- rephrase entirely. 'Some radiotracer and pharmaceutical information are always missing in the JSON sidecar files created from reading PET scanner data which is why dcm2niix4pet and ecat2nii take additional metadata. It is also possible to update existing JSON sidecar files with new metadata directly using updatejsonpetfile.m or update_json_pet_file.py. '

**Comment by CPernet** [comment]
does really work either -- changed to 'Those functions extend the image sidecar JSON file generated by dcm2niix with metadata from the users that are necessary to make files BIDS-compliant'

**Comment by nbeliy** [comment]
Check if it sounds better:

Those functions can extend the image sidecar JSON file generated by dcm2niix with user-provided metadata, making them BIDS-compliant.

Or:
In order to make image files generated by dcm2niix BIDS compliant, these function can extend image sidecar JSON file with user-provided metadata.

Or even:
Missing, but required by BIDS, metadata from image sidecar JSON file generated by dcm2niix can be completed by user.

**Comment by nbeliy** [comment]
What about the `further testing`

> The Python code was developed to mirror the Matlab code, further testing data reading (i.e., which file bits are read according to the PET data frames) and writing, relying here on Nibabel [@brett_nipynibabel_2023].

I propose:

The Python code for PET conversion mirrors the Matlab code, except writing of Nifti files, which are delegated to Nibabel [@brett_nipynibabel_2023].

Or something similar. You mention the tests of conversion directly after, so no need `further data testing`.

**Comment by CPernet** [comment]
done
